mumps 5.6.2 (new formula)

johnwparent opened the pull request:
<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

Currently the only tap that I could find vendoring MUMPS is no longer maintained and is out of date. Packages currently in brew/core actually depend on MUMPS and vendor the package as a resource rather than a package in its own right, which has resulted in some users installing things like `ipopt` simply to get the mumps headers/etc. 
MUMPS should be a first class package.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,72 @@
+  depends_on "gcc"
+  depends_on "openblas"
+
+  def install
```
There is so much magic here, can you try and upstream as much as possible?

johnwparent replied:
Are you referring to the custom installation logic? The upstream build system does not provide such a target. I can attempt to contribute one upstream (although MUMPS does not make this process obvious) but I'd rather not hold up this PR for MUMPs annual releases....

SMillerDev replied:
We can use patches if they're submitted upstream, but without that it'll be much harder to maintain this in the long run

johnwparent replied:
To be clear, the only "magic" we're discussing is the installation phase?

SMillerDev replied:
Yeah, setting a bunch of variables that could be set from upstream like: `optf = ["OPTF=-O"]` for example

johnwparent replied:
I've removed a number of these changes, including the one refered to above, but I'll respond w.r.t. to the additional comment make to this effect.

MUMPS has no public VCS that homebrew can link to to pull a patch. We're stuck waiting for a release to come out (and these changes would have to stay as is anyway), and then someone would have evaluate whether to keep these or not when the package is updated. 

I'll email a patch file to the upstream MUMPS mailing list, but I do not have the bandwidth to make a more pervasive update to their build system if they require a more thorough fix to the issues on top of the minimally functional patch.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,72 @@
+    make_args = ["RANLIB=echo", "CDEFS=-DAdd_"]
+    # floating point opts
+    optf = ["OPTF=-O"]
+    gcc_major_ver = Formula["gcc"].any_installed_version.major
+    optf << "-fallow-argument-mismatch" if gcc_major_ver >= 10
```
The current one is 13, so this doesn't have to be a conditional

johnwparent replied:
Forgive me, I'm not overly familiar with homebrew, but is there no scenario where a user could have an older GCC? Or is that a scenario that's just not covered?

SMillerDev replied:
You're depending on gcc, so it'll never be lower than it is now (gcc 13)

johnwparent replied:
Makes sense, thanks!

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,72 @@
+
+    system "make", "allshared", *make_args
+
+    so = OS.mac? ? "dylib" : "so"
+
+    lib.install Dir["lib/*.#{so}"]
+    lib.install "libseq/libmpiseq.#{so}"
```
This can use `shared_library`

johnwparent replied:
Heard, will amend.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,72 @@
+    doc.install Dir["doc/*.pdf"]
+    pkgshare.install "examples"
+
+    prefix.install "Makefile.inc"  # For the record.
+    File.open(prefix/"make_args.txt", "w") do |f|
+      f.puts(make_args.join(" "))  # Record options passed to make.
+    end
```
Why?

johnwparent replied:
Oh I saw this in a few other formulae (not core), thought it was useful for debugging and forgot I put it in. Total cruft, will remove.

johnwparent commented:
@SMillerDev I have addressed most comments.

The remaining point of uncertainty from the previous review is the "magic". I've removed all that I can while still allowing MUMPS to build for me. I've reached out to the MUMPS folks to see about upstreaming the install target, but AFAICT they have no public VCS so I just have the patch locally. I don't foresee the quick installation logic I have here presenting much of a maintenance issue, and on the off chance MUMPS modifies their installation tree in a way the recipe doesn't handle, it's a quick update. Maintaining the patch arbitrarily seems much harder and I'd rather not hold off this MR on a mailing list response.

All other build system intervention is fairly site-specific (or are specific instructions from the MUMPS build docs) and as such can't really be upstreamed in a generic fashion.

fahasch commented on Formula/i/ipopt.rb:
```diff
@@ -21,26 +21,7 @@ class Ipopt < Formula
-      end
-    end
-  end
+  depends_on "mumps"
```
`depends_on "mumps"` should be put before `depends_on "openblas"` as  mumps implies openblas.

johnwparent replied:
Amended and pushed.

fahasch replied:
Great. I am afraid that now compilation of ipopt fails with:
``` Error: inreplace failed
  Makefile.inc:
    expected replacement of "@rpath/" with "/opt/homebrew/opt/mumps/lib/"```
   Is the formula working on your local installation?

johnwparent replied:
Admittedly I only tested mumps as that's what I was focused on, I'll verify ipopt now.

johnwparent replied:
On further inspection, looks like it's actually `mumps` failing, not `ipopt`. I've been testing on Linux, looks like the failure is coming from a MacOS logical branch. I've pushed an update.

fahasch commented:
The errors in the compilation might be due to the fact that the present formula compiles with clang. Given the fact that it is a (mixture) of fortran and c, compiled with gfortran, the gcc compiler might be more appropriate.

johnwparent commented:
> The errors in the compilation might be due to the fact that the present formula compiles with clang. Given the fact that it is a (mixture) of fortran and c, compiled with gfortran, the gcc compiler might be more appropriate.

@fahasch Thanks for the insight, I was under the impression that this stanza: `depends_on "gcc"` would enforce that gcc is the compiler, rather than `clang`. AFAIK `mumps` cannot be compiled with the clang compiler (even with flang) so this is likely the cause of the issue as you said.

fahasch commented:
I'm not completely sure: it is true that `ENV.cc` should be `gcc`. To be absolutely sure you can add `fails_with :clang` after `depends_on gcc`.

I just tried your formula locally (on MacOS). It turns out that the problem is with the linker. The option `-soname` is not supported by ld. It should be replaced by `-install_name`. The relevant parameter might be `SONAME`. You find some ideas how to organize the linker in the formula for [scotch](https://github.com/Homebrew/homebrew-core/blob/master/Formula/s/scotch.rb).

fahasch commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,55 @@
+  def install
+    makefile = "Makefile.G95.SEQ"
+    cp "Make.inc/" + makefile, "Makefile.inc"
+    inreplace "Makefile.inc", "-soname", "-install_name" unless OS.linux?
```
For the version to compile under MacOS, you should include the line
` inreplace "Makefile.inc", ".so", ".dylib" unless OS.linux?`

johnwparent replied:
Ugh makefiles.

Thanks for pointing that out, good catch. It is compiling as it is now though, but is this the source of the rpath issue?

johnwparent replied:
This seems to have been it! Messed with it locally and can't replicate the CI failure anymore. 

Thanks!!

fahasch replied:
Good luck with the test. Compiling of  MUMPS is really a hassle. Even more there is a benefit from having it in homebrew.

johnwparent replied:
Mumps compiled fine, I made a mistake installing the headers.

johnwparent commented:
@fahasch I believe I have made all requested edits and the package seems to be working now.

fahasch commented:
Agreed! Let us see what @SMillerDev thinks about the formula.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+  fails_with :clang
+
+  def install
+    makefile = "Makefile.G95.SEQ"
```
Why not use the one that the ipopt resource used?

johnwparent replied:
I can switch if it's a problem for more modern gfortrans, both files worked for me locally.

johnwparent replied:
Actually, the non G95 might work a bit better for MacOS (I develop on Linux), let me see if I can remove some of the lines modifying the makefile with that one.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+  def install
+    makefile = "Makefile.G95.SEQ"
+    cp "Make.inc/" + makefile, "Makefile.inc"
+    inreplace "Makefile.inc", "-soname", "-install_name" unless OS.linux?
+    inreplace "Makefile.inc", ".so", ".dylib" unless OS.linux?
```
Could you try and submit this upstream?

johnwparent replied:
Upstream comments addressed above.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+    cp "Make.inc/" + makefile, "Makefile.inc"
+    inreplace "Makefile.inc", "-soname", "-install_name" unless OS.linux?
+    inreplace "Makefile.inc", ".so", ".dylib" unless OS.linux?
+    inreplace "Makefile.inc", "# RPATH_OPT = -Wl,-rpath,/path/to/MUMPS_x.y.z/lib/", "RPATH_OPT = -Wl,-rpath,#{lib}"
```
@carlocab do you have an opinion maybe?

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+    inreplace "Makefile.inc", ".so", ".dylib" unless OS.linux?
+    inreplace "Makefile.inc", "# RPATH_OPT = -Wl,-rpath,/path/to/MUMPS_x.y.z/lib/", "RPATH_OPT = -Wl,-rpath,#{lib}"
+    make_args = ["CDEFS=-DAdd_", "OPTF=-fallow-argument-mismatch"]
+    make_args += ["CC=#{ENV["CC"]} -fPIC",
```
```suggestion
    make_args += ["CC=#{ENV.cc} -fPIC",
```

johnwparent replied:
Addressed

johnwparent commented:
@SMillerDev I believe I have either responded to or addressed your concerns/comments and this PR is ready for another round of review feedback.

johnwparent commented:
@SMillerDev @fahasch Just wanted to circle back on this to keep things moving forward, please let me know what I can do to promote the progressions of this MR.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+  license "CECILL-C"
+
+  # Core dependencies
+  depends_on "gcc"
```
Does it depend on GCC at runtime?

johnwparent replied:
No, how do I express a build/link only dep?

johnwparent replied:
Thanks!

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+  def install
+    makefile = "Makefile.G95.SEQ"
+    cp "Make.inc/" + makefile, "Makefile.inc"
+    inreplace "Makefile.inc", "-soname", "-install_name" unless OS.linux?
+    inreplace "Makefile.inc", ".so", ".dylib" unless OS.linux?
+    inreplace "Makefile.inc", "# RPATH_OPT = -Wl,-rpath,/path/to/MUMPS_x.y.z/lib/", "RPATH_OPT = -Wl,-rpath,#{lib}"
```
Have these issues been reported/fixed upstream?

johnwparent replied:
I am discussing ln:16 and ln:17 with the MUMPS developers, ln:18 is not an issue, it's an expected and documented part of the MUMPS build workflow if you want to build w/ RPATHS.

MikeMcQuaid replied:
> it's an expected and documented part of the MUMPS build workflow if you want to build w/ RPATHS.

Can you link to the documentation then here please?

```suggestion
    # Required to build with RPATH: 
    # https://some.li/nk
    inreplace "Makefile.inc", "-soname", "-install_name" unless OS.linux?
    inreplace "Makefile.inc", ".so", ".dylib" unless OS.linux?
    inreplace "Makefile.inc", "# RPATH_OPT = -Wl,-rpath,/path/to/MUMPS_x.y.z/lib/", "RPATH_OPT = -Wl,-rpath,#{lib}"
```

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  test do
+    cd testpath do
```
Tests are always in the testpath

johnwparent replied:
That makes a lot of sense. Will remove.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+    inreplace "Makefile.inc", "-soname", "-install_name" unless OS.linux?
+    inreplace "Makefile.inc", ".so", ".dylib" unless OS.linux?
+    inreplace "Makefile.inc", "# RPATH_OPT = -Wl,-rpath,/path/to/MUMPS_x.y.z/lib/", "RPATH_OPT = -Wl,-rpath,#{lib}"
+    make_args = ["CDEFS=-DAdd_", "OPTF=-fallow-argument-mismatch"]
+    make_args += ["CC=#{ENV.cc} -fPIC",
+                  "FC=gfortran -fPIC -fopenmp",
+                  "FL=gfortran -fPIC -fopenmp"]
+    # Default lib args
+    # Note: MUMPS link cannot find LAPACK without these
+    # lines
+    blas_lib = "-L#{Formula["openblas"].opt_lib} -lopenblas"
+    make_args << "LIBBLAS=#{blas_lib}"
+    make_args << "LAPACK=#{blas_lib}"
```
Can you add a comment with the upstream issue report for this? Because it shouldn't need this much custom configuration to compile right?

johnwparent replied:
I find it pretty common that makefile systems rely on non-trivial user intervention to correctly setup the make variables for a given build environment. MUMPS seems to take that approach.
The `OPTF` and `CDEFS` args are actually specifically required by the MUMPS build instructions docs, as are the blas and lapack specs and most of the extra compile flags. From the MUMPS build docs: _"in most cases, Makefile.inc should be adapted to fit with your architecture, libraries and compilers"_ . I absolutely agree there are improvements than can/should be made to the MUMPS build system, and I have reached out to the MUMPS devs.

I have reached out to the MUMPS devs about an install target, and a MacOS specific makefile, as well as a few of the places I do think we should not have to specify a make argument. _However_ the MUMPS developers do not maintain as public facing issue tracker, VCS, or archive of any kind AFAICT (from googling and some communication on the general mailing list), so there unfortunately is nothing for you folks to monitor, track, or reference. I am communicating with them through their dev mailing list, and general mailing list, which is also how I will submit patches if and when there are any.  I understand you would like to avoid the added maintenance of arguments specified to a build system, so how would you like to proceed here given this context?
To be clear, I am happy to upstream or iterate on whatever is definitively necessary to upstream to make this PR work for homebrew, but with no tracking, I think perhaps this PR can move in tandem with my upstreams to MUMPS.
Maybe there's someone at homebrew I can CC on some of the MUMPS threads?

Additionally, all that said, the lack of `-lblas` in `Make.inc/Makefile.G95.SEQ` is almost certainly a bug, and I will include that in my iterations with the MUMPS devs.

MikeMcQuaid replied:
> I find it pretty common that makefile systems rely on non-trivial user intervention to correctly setup the make variables for a given build environment

No, it's really not that common if you look at the software inside of Homebrew.

SMillerDev commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,56 @@
+    ENV.deparallelize
+    system "make", "allshared", *make_args
+
+    # Makefile provides no install target, perform as needed install
+    # Install libraries
+    lib.install Dir["lib/#{shared_library("*")}"]
+    lib.install Dir["libseq/libmpiseq#{shared_library("*")}"]
+    # Install headers
+    libexec.install "include"
+    (libexec/"include").install Dir["libseq/*.h"]
+    include.install_symlink Dir[libexec/"include/*"]
+    # Install docs and examples
+    doc.install Dir["doc/*.pdf"]
+    pkgshare.install "examples"
```
Can we submit a patch upstream that adds an install target?

johnwparent replied:
I am discussing this with the MUMPS developers on their dev mailing list. AFAICT (and from a response on the general mailing list) there are 0 public facing resources/trackers for MUMPS development, and changes are submitted via that same mailing list, so there is nothing for homebrew to track/reference that would have any meaning.

SMillerDev commented:
If there is a mailing list archive link, I think that would be good to add as reference. Otherwise this conversation will have to do.

johnwparent commented:
> If there is a mailing list archive link, I think that would be good to add as reference. Otherwise this conversation will have to do.

@SMillerDev I've been looking for an archive for a while now to no results. There's clearly an option to access the archive in the MUMPS mailing list signup UI, but the option appears non functional/enabled. 
https://listes.ens-lyon.fr/sympa/subscribe/mumps-users for context, you can see the unhighlighted option for "archive" in the table to the left of the page.

At this point, it seems like this conversation is all we'll have to document this issue, so LGTM?

johnwparent commented:
@SMillerDev Last I understood these changes were ready for the upstream a few weeks ago, are there remaining blockers? What can we do to move this PR along and prevent stale-ness?
Any help is appreciated, and thanks for your review so far!

johnwparent commented:
CI failure is unrelated to these changes (IO timeout and other unrelated failures) and I cannot re-run the pipelines.

MikeMcQuaid commented on Formula/m/mumps.rb:
```diff
@@ -0,0 +1,54 @@
+  # Core dependencies
+  depends_on "gcc" => :build
+  depends_on "openblas"
+  fails_with :clang
```
Does this fail always and unconditionally with Clang? Why is that?

MikeMcQuaid reviewed (changes requested):
I'm willing to change my mind here but: vendoring looks like a much better option here:
- the building and installation of this software is a mess
- it doesn't work with Clang (the default macOS compiler for an age)
- it seems extremely likely to break or have essential parts missed in future upgrades

Sorry @johnwparent, I appreciate all the effort here and am willing to change my mind here if there's enough documentation added and linked to that this all becomes obvious but, as-is, this seems like a lot of work to maintain and much more messy than the previous vendoring approach.

ProfFan commented:
It does work with clang, compiling MUMPS on my machine with Clang absolutely works.

Also it seems that all these environment modifications are not necessary - here is my `Makefile.inc`:

```
SCOTCHDIR  = $(shell brew --prefix scotch)
ISCOTCH    = -I$(SCOTCHDIR)/include
LSCOTCH    = -L$(SCOTCHDIR)/lib -lptesmumps -lscotch -lptscotch -lptscotcherr -lptscotcherrexit -lscotcherrexit

LPORDDIR = $(topdir)/PORD/lib/
IPORD    = -I$(topdir)/PORD/include/
LPORD    = -L$(LPORDDIR) -lpord

LMETISDIR = $(shell brew --prefix metis)
IMETIS    = -I$(LMETISDIR)/include

LMETIS    = -L$(LMETISDIR)/lib -lmetis

ORDERINGSF = -Dscotch -Dmetis -Dpord -Dptscotch
ORDERINGSC  = $(ORDERINGSF)

LORDERINGS = $(LMETIS) $(LPORD) $(LSCOTCH)
IORDERINGSF = $(ISCOTCH)
IORDERINGSC = $(IMETIS) $(IPORD) $(ISCOTCH)

# PLAT : use it to add a default suffix to the generated libraries
PLAT    = 
# Library extension, + C and Fortran "-o" option
# may be different under Windows
LIBEXT  = .a
OUTC    = -o 
OUTF    = -o 
# RM : remove files
RM      = /bin/rm -f
# CC : C compiler
CC      = cc
# FC : Fortran 90 compiler
FC      = mpif90
# FL : Fortran linker
FL      = mpif90
# AR : Archive object in a library
#      keep a space at the end if options have to be separated from lib name
AR      = ar vr 
# RANLIB : generate index of an archive file
#   (optionnal use "RANLIB = echo" in case of problem)
RANLIB  = ranlib
#RANLIB  = echo

# DEFINE HERE YOUR LAPACK LIBRARY
LAPACK = -llapack

# SCALAP should define the SCALAPACK and  BLACS libraries.
SCALAP  = -L$(shell brew --prefix scalapack)/lib -lscalapack

# INCLUDE DIRECTORY FOR MPI
INCPAR  = -I$(shell brew --prefix open-mpi)/include

# LIBRARIES USED BY THE PARALLEL VERSION OF MUMPS: $(SCALAP) and MPI
LIBPAR  = $(SCALAP) $(LAPACK) -L/usr/lib -lmpi

# The parallel version is not concerned by the next two lines.
# They are related to the sequential library provided by MUMPS,
# to use instead of ScaLAPACK and MPI.
INCSEQ  = -I$(topdir)/libseq
LIBSEQ  = $(LAPACK) -L$(topdir)/libseq -lmpiseq

# DEFINE HERE YOUR BLAS LIBRARY
LIBBLAS = -lblas

# DEFINE HERE YOUR PTHREAD LIBRARY
LIBOTHERS = -lpthread

# FORTRAN/C COMPATIBILITY:
#  Use:
#    -DAdd_ if your Fortran compiler adds an underscore at the end
#              of symbols,
#     -DAdd__ if your Fortran compiler adds 2 underscores,
#
#     -DUPPER if your Fortran compiler uses uppercase symbols
#
#     leave empty if your Fortran compiler does not change the symbols.
#

CDEFS = -DAdd_

#COMPILER OPTIONS
OPTF    = -O -fallow-argument-mismatch
OPTC    = -O -I.
OPTL    = -O

# CHOOSE BETWEEN USING THE SEQUENTIAL OR THE PARALLEL VERSION.

#Sequential:
#INCS = $(INCSEQ)
#LIBS = $(LIBSEQ)
#LIBSEQNEEDED = libseqneeded

#Parallel:
INCS = $(INCPAR)
LIBS = $(LIBPAR)
LIBSEQNEEDED = 
```

Works within a regular shell and just `make all`.
